client/js/spotify-gdpr.js

```diff
@@ -3,214 +3,106 @@
+      collatedRecords.intervals[i - ROLLING_WINDOW_SIZE].totals :
+      null;
+
+    for (const [artist, tracks] of entering) {
+      for (const [track, added] of tracks) {
+        const key = `${track} - ${artist}`;
+        const current = rollingSums.get(key) || 0;
+        rollingSums.set(key, current + added);
       }
     }
 
-    // take the current ordering of the songs and store it in the historical
-    // index list
-    for (const [id, index] of indices.entries()) {
-      let indexHistory;
-      if (indicesHistory.has(id)) {
-        indexHistory = indicesHistory.get(id);
-        indexHistory.push(index);
-      } else {
-        // this song wasn't in the record, backfill all of the indices
-        indexHistory = [];
-
-        for (let i = 0; i < generation; i++) {
-          indexHistory.push(null);
+    if (exiting) {
+      for (const [artist, tracks] of exiting) {
+        for (const [track, removed] of tracks) {
+          const key = `${track} - ${artist}`;
+          const current = rollingSums.get(key) || 0;
+          rollingSums.set(key, current - removed);
         }
-
-        indexHistory.push(index);
-        indicesHistory.set(id, indexHistory);
       }
     }
 
-    xAxis.push(new Date(+xAxis[xAxis.length - 1] - MS_PER_DAY));
-    generation++;
+    // turn into array of entries
+    [...rollingSums]
+        // sort by play time in the last 28 days
+        .sort((entryA, entryB) => {
+          if (entryA[1] > entryB[1]) return -1;
+          if (entryA[1] < entryB[1]) return 1;
+          return 0;
+        })
+        // select top 50
+        .slice(0, 15)
+        // add indexes to track history
+        .forEach((entry, index) => {
+          const [key] = entry;
+
+          let trackHistory = rankingHistory.get(key);
+
+          if (!trackHistory) {
+            trackHistory = [];
+            rankingHistory.set(key, trackHistory);
+          }
+
+          while (trackHistory.length < i) {
+            // if the track was missing from the chart, add null values
+            trackHistory.push(null);
+          }
+
+          trackHistory.push(index);
+        });
   }
 
   // format of a chart series in c3 is ['name of series', point1, point2, ...]
-  const chartData =
-    /** @type {Array<[string, ...number[]]>} */
-    (Array
-        .from(indicesHistory.entries())
-        .map(([id, history]) => [id, ...history]));
-
-  chart.load({
-    columns: [['date', ...xAxis], ...chartData],
-  });
+  const chartData = Array
+      .from(rankingHistory.entries())
+      .map(([id, history]) => [id, ...history]);
+
+  // load data in segments to avoid hanging the browser
+  for (let i = 0; i < chartData.length; i++) {
+    const chartDataSegment = chartData.slice(i, i + 15);
+    await new Promise((resolve) => chart.load({
+      columns: [
+        ['date', ...rankingDates],
+        ...chartDataSegment,
+      ],
+      done: resolve,
+    }));
+    await sleep(400);
```

Comment:
Why the sleep, and why 400ms?

Reply:
> `// load data in segments to avoid hanging the browser`

JS execution is single-threaded, so there's no way for me to do such an expensive operation without causing the window to hang for tens of seconds. The `sleep` yields control to the browser temporarily to make things remain responsive.

400ms is the Doherty Threshold, but I could've picked any number less than 500 I suppose, even 0.

Reply:
> 400ms is the Doherty Threshold
That was a fun read.

Have you experienced hanging when you don't do this? I'm pretty sure you're already yielding the rendering thread when you `await chart.load(...)` since your zip library uses webworkers (iirc). If you do experience hanging, can you add a comment that the 15 segments is meant to be roughly 400ms worth of work, so you're doing 400ms on and 400ms off. Otherwise it seems like the 400ms is there to allow you to wait on some specific task to be completed.

Reply:
The zip library uses web workers, but at this point, the zip library is no longer involved (the data has already been loaded and is now being stuffed into the chart). I did experience extreme hanging when I didn't do this, to the point where it was kind of unusable.

Reply:
SGTM. Yeah, just document roughly how long the chunks hang, and try get that number under 400ms.